### PR TITLE
Improved error message on nested proofs

### PR DESCRIPTION
**Pull request by fakusb**
Improved error message on nested proofs to include most common reason when this happens on accident.

Most users simply forgot to close a proof and don't want nested proofs.

**Kind:** documentation.

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).

Is a changelog entry needed for such things?

### CODE REVIEW

**Comment by gares** [comment]
Great initiative.

**Comment by gares** [comment]
@coq/doc-maintainers this fix is very conservative. What do you think to put upfront the second part of the message?
It would give something like:
> Nested proofs are not allowed, you probably forgot to close the last `Proof.` with `Qed.` or `Defined.`.  Coq can be forced to tolerate nesting by setting the Nested Proofs Allowed flag on.

**Comment by fakusb** [comment]
I removed the "changelog" todo item in this PR as i now understand that it is only needed when the "real" behaviour of cow changes, not only the content of error messages

**Comment by gares** [comment]
(CI is stuck, I'll run it once we have feedback by the doc maintainers)

**Comment by Zimmi48** [comment]
> @coq/doc-maintainers this fix is very conservative. What do you think to put upfront the second part of the message?
> It would give something like:
> 
> > Nested proofs are not allowed, you probably forgot to close the last `Proof.` with `Qed.` or `Defined.`.  Coq can be forced to tolerate nesting by setting the Nested Proofs Allowed flag on.

Some users will get mad if they read "Nested proofs are not allowed", so what about:

>Nested proofs are not allowed by default, and discouraged for beginners. This error probably means that you forgot to close ... If you really intended to use nested proofs, you can do so by turning the Nested Proofs Allowed flag on.

**Comment by Zimmi48** [comment]
BTW, @gares. I personally used to consider that user message improvements are backportable even between patch-level releases (we don't guarantee output message compatibility). But that's your call!

**Comment by fakusb** [comment]
how can I render the dune-doc locally to check that my adition is correct?

File: stm/stm.ml
Comment (jfehrle) [inline stm/stm.ml:2684]: Great idea to improve error messages.

This probably should use double quote (`"`) instead of a back-tick or grave (``` ` ```).  Sphinx doesn't interpret back ticks when they appear in error messages, BTW.

Also, beyond the scope of this PR, the error message index is a bit of a mess.  The entries are not wrapped (though you can jump to the text to see the full text) and if I click on "n", it becomes unreadable (there's no way to scroll left).

image

image

**Comment by fakusb** [comment]
I improved on the message by putting the stuff beginners need to see first, as @Zimmi48  mentioned, and improved the formatting of the source.

File: stm/stm.ml
Comment (Zimmi48) [reply stm/stm.ml]: > Also, beyond the scope of this PR, the error message index is a bit of a mess. The entries are not wrapped (though you can jump to the text to see the full text) and if I click on "n", it becomes unreadable (there's no way to scroll left).

Yes, that's worth opening an issue IMHO.

File: doc/sphinx/language/core/definitions.rst
Comment (Zimmi48) [inline doc/sphinx/language/core/definitions.rst:176]:
```suggestion
   .. exn:: Nested proofs are discouraged and not allowed by default. This error probably means that you forgot to close the last "Proof." with "Qed." or "Defined.". \
```

File: stm/stm.ml
Comment (Zimmi48) [inline stm/stm.ml:2685]:
```suggestion
           to close the last \"Proof.\" with \"Qed.\" or \"Defined.\". \
```

File: stm/stm.ml
Comment (Zimmi48) [inline stm/stm.ml:2687]: Should this become a `strbrk`?

Comment (fakusb) [reply stm/stm.ml]: I changed the three things you mentioned. Whats the difference? (the comment above strbrk says "to remove" btw)

Comment (gares) [reply stm/stm.ml]: IIRC strbrk consider each space a potential breaking point, while str should be printed all on the same line

Comment (gares) [reply stm/stm.ml]: Oh, and just for your info, in recent ocaml you can `{| multi line string with no need to escape " |}`

Comment (gares) [reply stm/stm.ml]: (but here indentation may get in the way...)

**Review by Zimmi48 — APPROVED** [review]

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**
